src/Microsoft.AspNetCore.Rewrite/Internal/IISUrlRewrite/UrlRewriteFileParser.cs

```diff
@@ -138,61 +138,69 @@ private void ParseCondition(XElement condition, UrlRewriteRuleBuilder builder, P
+                throw new UrlRewriteParseException(condition, formatException.Message, formatException);
             }
         }
 
         private void ParseUrlAction(XElement urlAction, UrlRewriteRuleBuilder builder, bool stopProcessing)
         {
             var actionType = ParseEnum(urlAction, RewriteTags.Type, ActionType.None);
-            var redirectType = ParseEnum(urlAction, RewriteTags.RedirectType, RedirectType.Permanent);
-            var appendQuery = ParseBool(urlAction, RewriteTags.AppendQueryString, defaultValue: true);
 
-            string url = string.Empty;
-            if (urlAction.Attribute(RewriteTags.Url) != null)
+            try
             {
-                url = urlAction.Attribute(RewriteTags.Url).Value;
-                if (string.IsNullOrEmpty(url))
+                UrlAction action;
+                switch (actionType)
                 {
-                    ThrowUrlFormatException(urlAction, "Url attribute cannot contain an empty string");
+                    case ActionType.None:
+                        action = new VoidAction(stopProcessing ? RuleResult.SkipRemainingRules : RuleResult.ContinueRules);
+                        break;
+                    case ActionType.Rewrite:
+                    case ActionType.Redirect:
+                        var url = string.Empty;
+                        if (urlAction.Attribute(RewriteTags.Url) != null)
+                        {
+                            url = urlAction.Attribute(RewriteTags.Url).Value;
+                            if (string.IsNullOrEmpty(url))
+                            {
+                                throw new UrlRewriteParseException(urlAction, "Url attribute cannot contain an empty string");
+                            }
+                        }
+
+                        var urlPattern = _inputParser.ParseInputString(url);
+                        var appendQuery = ParseBool(urlAction, RewriteTags.AppendQueryString, defaultValue: true);
+
+                        if (actionType == ActionType.Rewrite)
+                        {
+                            action = new RewriteAction(stopProcessing ? RuleResult.SkipRemainingRules : RuleResult.ContinueRules, urlPattern, appendQuery);
+                        }
+                        else
+                        {
+                            var redirectType = ParseEnum(urlAction, RewriteTags.RedirectType, RedirectType.Permanent);
+                            action = new RedirectAction((int)redirectType, urlPattern, appendQuery);
+                        }
+                        break;
+                    case ActionType.AbortRequest:
+                        action = new AbortAction();
+                        break;
+                    case ActionType.CustomResponse:
+                        int statusCode;
+                        if (!int.TryParse(urlAction.Attribute(RewriteTags.StatusCode)?.Value, out statusCode))
```

Comment:
On second thought, I think this should actually throw if there is no status code provided. Its required in the IIS Url rewrite module.
Thoughts @Tratcher ?